NATS-Jetstream is losing scheduled messages in combination with stream-sourcing and workqueue retention

### Observed behavior

Hello together,
I stumbled over the issue that in certain configurations, NATS-Jetstream in combination with the Message-Scheduling Feature, Stream-Sourcing and Workqueue-Retention is quietly losing/dropping scheduled messages.

The setup is the following:
It is a nats-cluster with multiple instances (e.g. 3)
Two streams exist: 
- First one is a Stream with scheduled messages (allows Msg-Schedules), that stream is using workqueue retention and also has multiple replicas 
- Second Stream with default settings that is sourcing the schedule-target subject of the messages in the first stream


If you now schedule multiple messages for the future (eg 4), and the message that was sent last has the most recent schedule_at date, **only the most recent message gets scheduled and the other three just dissapear**.
**The problem only occurs with multiple replicas in the scheduled-stream, if the replica is set to one, all the messages appear in the second (sourced) stream.**


### Expected behaviour
Only the last message with the most recent schedule_at date appears in the second-stream, the other messages are silently getting lost

```
=== RUN   TestWorkQueueSourceMustNotRemoveLaterSchedules
    repro_test.go:71: Published "x.latest", scheduled for 2026-07-27T15:15:15Z
    repro_test.go:71: Published "x.later-2", scheduled for 2026-07-27T15:15:13Z
    repro_test.go:71: Published "x.later-1", scheduled for 2026-07-27T15:15:11Z
    repro_test.go:71: Published "x.early", scheduled for 2026-07-27T15:15:09Z
    repro_test.go:84: SCHEDULES messages=0 first_seq=6 last_seq=5
    repro_test.go:90: TARGET messages=1 first_seq=1 last_seq=1
    repro_test.go:97: BUG REPRODUCED: TARGET messages=1, want 4
--- FAIL: TestWorkQueueSourceMustNotRemoveLaterSchedules (13.00s)
```

### Expected behavior

All 4 messages appear at the correct date in the schedule-target and are then getting sourced into the second stream
```
=== RUN   TestWorkQueueSourceMustNotRemoveLaterSchedules
    repro_test.go:71: Published "x.latest", scheduled for 2026-07-27T15:16:00Z
    repro_test.go:71: Published "x.later-2", scheduled for 2026-07-27T15:15:58Z
    repro_test.go:71: Published "x.later-1", scheduled for 2026-07-27T15:15:56Z
    repro_test.go:71: Published "x.early", scheduled for 2026-07-27T15:15:54Z
    repro_test.go:84: SCHEDULES messages=0 first_seq=9 last_seq=8
    repro_test.go:90: TARGET messages=4 first_seq=1 last_seq=4
--- PASS: TestWorkQueueSourceMustNotRemoveLaterSchedules (12.82s)
```

### Server and client version

github.com/nats-io/nats-server/v2 v2.14.3
github.com/nats-io/nats.go v1.52.0

### Host environment

_No response_

### Steps to reproduce

**Can be easily reproduced with the following example text-case:**
```go
package repro

import (
	"context"
	"fmt"
	"net"
	"path/filepath"
	"testing"
	"time"

	natsserver "github.com/nats-io/nats-server/v2/server"
	"github.com/nats-io/nats.go"
	"github.com/nats-io/nats.go/jetstream"
)

func TestWorkQueueSourceMustNotRemoveLaterSchedules(t *testing.T) {
	servers := startCluster(t)
	connection, err := nats.Connect(servers[0].ClientURL())
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(connection.Close)

	js, err := jetstream.New(connection)
	if err != nil {
		t.Fatal(err)
	}
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	schedules, err := js.CreateStream(ctx, jetstream.StreamConfig{
		Name:              "SCHEDULES",
		Subjects:          []string{"x.>"},
		Retention:         jetstream.WorkQueuePolicy,
		Replicas:          2,
		AllowMsgSchedules: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	target, err := js.CreateStream(ctx, jetstream.StreamConfig{
		Name: "TARGET",
		Sources: []*jetstream.StreamSource{{
			Name:          "SCHEDULES",
			FilterSubject: "x.target",
		}},
	})
	if err != nil {
		t.Fatal(err)
	}

	firstDue := time.Now().UTC().Truncate(time.Second).Add(time.Second)
	for _, schedule := range []struct {
		subject string
		at      time.Time
	}{
		{subject: "x.latest", at: firstDue.Add(6 * time.Second)},
		{subject: "x.later-2", at: firstDue.Add(4 * time.Second)},
		{subject: "x.later-1", at: firstDue.Add(2 * time.Second)},
		{subject: "x.early", at: firstDue},
	} {
		if _, err := js.Publish(
			ctx,
			schedule.subject,
			[]byte(schedule.subject),
			jetstream.WithScheduleAt(schedule.at),
			jetstream.WithScheduleTarget("x.target"),
		); err != nil {
			t.Fatal(err)
		}
		t.Logf("Published %q, scheduled for %s", schedule.subject, schedule.at.Format(time.RFC3339))
	}

	time.Sleep(time.Until(firstDue.Add(9 * time.Second)))

	scheduleInfo, err := schedules.Info(ctx)
	if err != nil {
		t.Fatal(err)
	}
	targetInfo, err := target.Info(ctx)
	if err != nil {
		t.Fatal(err)
	}
	t.Logf(
		"SCHEDULES messages=%d first_seq=%d last_seq=%d",
		scheduleInfo.State.Msgs,
		scheduleInfo.State.FirstSeq,
		scheduleInfo.State.LastSeq,
	)
	t.Logf(
		"TARGET messages=%d first_seq=%d last_seq=%d",
		targetInfo.State.Msgs,
		targetInfo.State.FirstSeq,
		targetInfo.State.LastSeq,
	)
	if targetInfo.State.Msgs != 4 {
		t.Fatalf("BUG REPRODUCED: TARGET messages=%d, want 4", targetInfo.State.Msgs)
	}
}

func startCluster(t *testing.T) []*natsserver.Server {
	t.Helper()
	ports := freePorts(t, 3)
	routes := fmt.Sprintf(
		"nats-route://127.0.0.1:%d,nats-route://127.0.0.1:%d,nats-route://127.0.0.1:%d",
		ports[0],
		ports[1],
		ports[2],
	)

	servers := make([]*natsserver.Server, 0, len(ports))
	for index, port := range ports {
		server, err := natsserver.NewServer(&natsserver.Options{
			ServerName:        fmt.Sprintf("repro-%d", index+1),
			Host:              "127.0.0.1",
			Port:              natsserver.RANDOM_PORT,
			JetStream:         true,
			JetStreamMaxStore: 256 << 20,
			StoreDir:          filepath.Join(t.TempDir(), fmt.Sprintf("repro-%d", index+1)),
			NoLog:             true,
			NoSigs:            true,
			Cluster: natsserver.ClusterOpts{
				Name: "repro",
				Host: "127.0.0.1",
				Port: port,
			},
			Routes: natsserver.RoutesFromStr(routes),
		})
		if err != nil {
			t.Fatal(err)
		}
		server.Start()
		if !server.ReadyForConnections(5 * time.Second) {
			t.Fatalf("server %d not ready", index+1)
		}
		servers = append(servers, server)
	}
	t.Cleanup(func() {
		for _, server := range servers {
			server.Shutdown()
		}
		for _, server := range servers {
			server.WaitForShutdown()
		}
	})

	time.Sleep(3 * time.Second)
	return servers
}

func freePorts(t *testing.T, count int) []int {
	t.Helper()
	listeners := make([]net.Listener, 0, count)
	for range count {
		listener, err := net.Listen("tcp", "127.0.0.1:0")
		if err != nil {
			t.Fatal(err)
		}
		listeners = append(listeners, listener)
	}

	ports := make([]int, 0, count)
	for _, listener := range listeners {
		ports = append(ports, listener.Addr().(*net.TCPAddr).Port)
		if err := listener.Close(); err != nil {
			t.Fatal(err)
		}
	}
	return ports
}
```

--- a/server/consumer.go
+++ b/server/consumer.go
@@ -3688,7 +3688,7 @@ func (o *consumer) processAckMsgLocked(sseq, dseq, dc uint64, reply string, doSa
 	// violate lock ordering with respect to the stream.
 	ackInPlace := o.node == nil && o.retention != LimitsPolicy && needLock
 
-	var sgap, floor uint64
+	var ackAllSeqs []uint64
 	var needSignal bool
 
 	switch o.cfg.AckPolicy {
@@ -3717,17 +3717,22 @@ func (o *consumer) processAckMsgLocked(sseq, dseq, dc uint64, reply string, doSa
 		if o.maxp > 0 && len(o.pending) >= o.maxp {
 			needSignal = true
 		}
-		sgap = sseq - o.asflr
-		floor = sseq // start at same and set lower as we go.
+		sgap := sseq - o.asflr
 		o.adflr, o.asflr = dseq, sseq
 
+		// At most the pending entries below the ack, don't over-allocate.
+		ackAllSeqs = make([]uint64, 0, min(uint64(len(o.pending)), sgap-1))
+
 		remove := func(seq uint64) {
+			// Only collect what was actually delivered to us.
+			if ackInPlace && seq != sseq {
+				if _, ok := o.pending[seq]; ok {
+					ackAllSeqs = append(ackAllSeqs, seq)
+				}
+			}
 			delete(o.pending, seq)
 			delete(o.rdc, seq)
 			o.removeFromRedeliverQueue(seq)
-			if seq < floor {
-				floor = seq
-			}
 		}
 		// Determine if smarter to walk all of pending vs the sequence range.
 		if sgap > uint64(len(o.pending)) {
@@ -3757,13 +3762,9 @@ func (o *consumer) processAckMsgLocked(sseq, dseq, dc uint64, reply string, doSa
 	unlock()
 
 	if ackInPlace {
-		if sgap > 1 {
-			// FIXME(dlc) - This can very inefficient, will need to fix.
-			for seq := sseq; seq >= floor; seq-- {
-				mset.ackMsg(o, seq)
-			}
-		} else {
-			mset.ackMsg(o, sseq)
+		mset.ackMsg(o, sseq)
+		for _, seq := range ackAllSeqs {
+			mset.ackMsg(o, seq)
 		}
 	}
 

--- a/server/jetstream_cluster.go
+++ b/server/jetstream_cluster.go
@@ -7128,13 +7128,22 @@ func (o *consumer) processReplicatedAck(dseq, sseq uint64) error {
 	// Update activity.
 	o.lat = time.Now()
 
-	var sagap uint64
-	if o.cfg.AckPolicy == AckAll || o.cfg.AckPolicy == AckFlowControl {
+	var ackAllSeqs []uint64
+	if o.retention != LimitsPolicy && (o.cfg.AckPolicy == AckAll || o.cfg.AckPolicy == AckFlowControl) {
 		// Always use the store state, as o.asflr is skipped ahead already.
-		// Capture before updating store.
+		// Capture before updating store, which clears the pending below.
 		state, err := o.store.BorrowState()
 		if err == nil {
-			sagap = sseq - state.AckFloor.Stream
+			// Only need to collect if the ack covers more than the sequence itself.
+			if sagap := sseq - state.AckFloor.Stream; sagap > 1 {
+				// At most the pending entries below the ack, don't over-allocate.
+				ackAllSeqs = make([]uint64, 0, min(uint64(len(state.Pending)), sagap-1))
+				for seq := range state.Pending {
+					if seq < sseq {
+						ackAllSeqs = append(ackAllSeqs, seq)
+					}
+				}
+			}
 		}
 	}
 
@@ -7164,13 +7173,9 @@ func (o *consumer) processReplicatedAck(dseq, sseq uint64) error {
 	}
 	o.mu.Unlock()
 
-	if sagap > 1 {
-		// FIXME(dlc) - This is very inefficient, will need to fix.
-		for seq := sseq; seq > sseq-sagap; seq-- {
-			mset.ackMsg(o, seq)
-		}
-	} else {
-		mset.ackMsg(o, sseq)
+	mset.ackMsg(o, sseq)
+	for _, seq := range ackAllSeqs {
+		mset.ackMsg(o, seq)
 	}
 	return nil
 }

--- a/server/jetstream_cluster_3_test.go
+++ b/server/jetstream_cluster_3_test.go
@@ -11522,3 +11522,246 @@ func TestJetStreamClusterMetaSnapshotRecoveryScaleConsumer(t *testing.T) {
 		return nil
 	})
 }
+
+// https://github.com/nats-io/nats-server/issues/8423
+func TestJetStreamClusterWorkQueueSourceMustNotAckOtherSubjects(t *testing.T) {
+	for _, replicas := range []int{1, 3} {
+		t.Run(fmt.Sprintf("R%d", replicas), func(t *testing.T) {
+			c := createJetStreamClusterExplicit(t, "R3S", 3)
+			defer c.shutdown()
+
+			nc, js := jsClientConnect(t, c.randomServer())
+			defer nc.Close()
+
+			_, err := jsStreamCreate(t, nc, &StreamConfig{
+				Name:      "WQ",
+				Subjects:  []string{"x.>"},
+				Storage:   FileStorage,
+				Retention: WorkQueuePolicy,
+				Replicas:  replicas,
+			})
+			require_NoError(t, err)
+
+			// Wrap two matching sourced messages around a gap.
+			_, err = js.Publish("x.target", nil)
+			require_NoError(t, err)
+			for range 3 {
+				_, err = js.Publish("x.other", nil)
+				require_NoError(t, err)
+			}
+			_, err = js.Publish("x.target", nil)
+			require_NoError(t, err)
+
+			_, err = jsStreamCreate(t, nc, &StreamConfig{
+				Name:     "TARGET",
+				Storage:  FileStorage,
+				Replicas: replicas,
+				Sources:  []*StreamSource{{Name: "WQ", FilterSubject: "x.target"}},
+			})
+			require_NoError(t, err)
+
+			// Wait for both messages to be sourced.
+			checkFor(t, 5*time.Second, 100*time.Millisecond, func() error {
+				si, err := js.StreamInfo("TARGET")
+				if err != nil {
+					return err
+				}
+				if si.State.Msgs != 2 {
+					return fmt.Errorf("expected 2 messages in TARGET, got %d", si.State.Msgs)
+				}
+				return nil
+			})
+
+			// The ack of sequence 5 must not have removed sequences 2-4 on "x.other".
+			checkFor(t, 2*time.Second, 200*time.Millisecond, func() error {
+				si, err := js.StreamInfo("WQ")
+				if err != nil {
+					return err
+				}
+				if si.State.Msgs != 3 {
+					return fmt.Errorf("expected 3 messages in WQ, got %d (first=%d, last=%d)",
+						si.State.Msgs, si.State.FirstSeq, si.State.LastSeq)
+				}
+				return nil
+			})
+		})
+	}
+}
+
+// https://github.com/nats-io/nats-server/issues/8423
+func TestJetStreamClusterScheduledMessagesWithWorkQueueSource(t *testing.T) {
+	for _, replicas := range []int{1, 3} {
+		t.Run(fmt.Sprintf("R%d", replicas), func(t *testing.T) {
+			c := createJetStreamClusterExplicit(t, "R3S", 3)
+			defer c.shutdown()
+
+			nc, js := jsClientConnect(t, c.randomServer())
+			defer nc.Close()
+
+			_, err := jsStreamCreate(t, nc, &StreamConfig{
+				Name:              "SCHEDULES",
+				Subjects:          []string{"x.>"},
+				Storage:           FileStorage,
+				Retention:         WorkQueuePolicy,
+				Replicas:          replicas,
+				AllowMsgSchedules: true,
+			})
+			require_NoError(t, err)
+
+			publishSchedule := func(subject string, due time.Time) {
+				t.Helper()
+				m := nats.NewMsg(subject)
+				m.Header.Set(JSSchedulePattern, "@at "+due.UTC().Format(time.RFC3339))
+				m.Header.Set(JSScheduleTarget, "x.target")
+				_, err = js.PublishMsg(m)
+				require_NoError(t, err)
+			}
+			// A fired schedule writes to "x.target" and purges its own message, so we
+			// can wait on the resulting stream sequence.
+			waitForFire := func(lseq uint64) {
+				t.Helper()
+				checkFor(t, 5*time.Second, 50*time.Millisecond, func() error {
+					si, err := js.StreamInfo("SCHEDULES")
+					if err != nil {
+						return err
+					}
+					if si.State.LastSeq != lseq {
+						return fmt.Errorf("expected last sequence %d, got %d", lseq, si.State.LastSeq)
+					}
+					return nil
+				})
+			}
+
+			soon := time.Now().Add(time.Second)
+			never := time.Now().Add(time.Hour)
+
+			// Fire a schedule first, so a message for "x.target" ends up below the
+			// schedules published next. Without a match at either end the sourcing
+			// consumer has one pending sequence, and the ack range collapses down to
+			// it, hiding the bug.
+			publishSchedule("x.fire-1", soon) // Sequence 1, purged once fired.
+			waitForFire(2)                    // Sequence 2, for "x.target".
+
+			// These are due long after the test, they must be kept.
+			publishSchedule("x.keep-1", never) // Sequence 3.
+			publishSchedule("x.keep-2", never) // Sequence 4.
+
+			publishSchedule("x.fire-2", soon) // Sequence 5, purged once fired.
+			waitForFire(6)                    // Sequence 6, for "x.target".
+
+			// Only source once both messages for "x.target" are there, so a single
+			// AckAll spans the schedules in between.
+			_, err = jsStreamCreate(t, nc, &StreamConfig{
+				Name:     "TARGET",
+				Storage:  FileStorage,
+				Replicas: replicas,
+				Sources:  []*StreamSource{{Name: "SCHEDULES", FilterSubject: "x.target"}},
+			})
+			require_NoError(t, err)
+
+			// Both fired schedules must end up in TARGET.
+			checkFor(t, 5*time.Second, 100*time.Millisecond, func() error {
+				si, err := js.StreamInfo("TARGET")
+				if err != nil {
+					return err
+				}
+				if si.State.Msgs != 2 {
+					return fmt.Errorf("expected 2 messages in TARGET, got %d", si.State.Msgs)
+				}
+				return nil
+			})
+
+			// The ack of sequence 6 must not have removed the schedules at 3 and 4.
+			checkFor(t, 2*time.Second, 200*time.Millisecond, func() error {
+				si, err := js.StreamInfo("SCHEDULES")
+				if err != nil {
+					return err
+				}
+				if si.State.Msgs != 2 {
+					return fmt.Errorf("expected 2 schedules in SCHEDULES, got %d (first=%d, last=%d)",
+						si.State.Msgs, si.State.FirstSeq, si.State.LastSeq)
+				}
+				return nil
+			})
+		})
+	}
+}
+
+func TestJetStreamClusterAckAllAfterConsumerFilterUpdate(t *testing.T) {
+	c := createJetStreamClusterExplicit(t, "R3S", 3)
+	defer c.shutdown()
+
+	nc, js := jsClientConnect(t, c.randomServer())
+	defer nc.Close()
+
+	_, err := jsStreamCreate(t, nc, &StreamConfig{
+		Name:      "WQ",
+		Subjects:  []string{"x.>"},
+		Storage:   FileStorage,
+		Retention: WorkQueuePolicy,
+		Replicas:  3,
+	})
+	require_NoError(t, err)
+
+	_, err = js.Publish("x.a", nil)
+	require_NoError(t, err)
+	_, err = js.Publish("x.b", nil)
+	require_NoError(t, err)
+
+	inbox := nats.NewInbox()
+	sub, err := nc.SubscribeSync(inbox)
+	require_NoError(t, err)
+	defer sub.Unsubscribe()
+
+	// A flow-controlled consumer acks like an AckAll, and is allowed on a WQ stream.
+	cfg := ConsumerConfig{
+		Durable:        "C",
+		DeliverSubject: inbox,
+		FilterSubjects: []string{"x.a", "x.b"},
+		AckPolicy:      AckFlowControl,
+		FlowControl:    true,
+		Heartbeat:      time.Second,
+	}
+	_, err = jsConsumerCreate(t, nc, "WQ", cfg, false)
+	require_NoError(t, err)
+
+	// Both messages must be delivered, so both are pending. Also need the flow
+	// control subject, replying to that is what acks for this consumer.
+	var delivered int
+	var fcSubject string
+	for delivered < 2 || fcSubject == _EMPTY_ {
+		m, err := sub.NextMsg(2 * time.Second)
+		require_NoError(t, err)
+		if len(m.Header) == 0 {
+			delivered++
+		} else if m.Reply != _EMPTY_ {
+			fcSubject = m.Reply
+		} else if stalled := m.Header.Get(JSConsumerStalled); stalled != _EMPTY_ {
+			fcSubject = stalled
+		}
+	}
+
+	// Narrow the filters, "x.a" is pending but no longer matches.
+	cfg.FilterSubjects = []string{"x.b"}
+	_, err = jsConsumerCreate(t, nc, "WQ", cfg, false)
+	require_NoError(t, err)
+
+	// Reply to flow control, acking everything delivered up to sequence 2.
+	m := nats.NewMsg(fcSubject)
+	m.Header.Set(JSLastConsumerSeq, "2")
+	m.Header.Set(JSLastStreamSeq, "2")
+	require_NoError(t, nc.PublishMsg(m))
+
+	// Both messages were acked by the consumer, so both must be removed.
+	checkFor(t, 5*time.Second, 100*time.Millisecond, func() error {
+		si, err := js.StreamInfo("WQ")
+		if err != nil {
+			return err
+		}
+		if si.State.Msgs != 0 {
+			return fmt.Errorf("expected 0 messages in WQ, got %d (first=%d, last=%d)",
+				si.State.Msgs, si.State.FirstSeq, si.State.LastSeq)
+		}
+		return nil
+	})
+}

